Obsfucate the LDAP system password in toString and debug.log

craigtaverner opened the pull request:
changelog: Fixed visibility of LDAP system password in debug.log through new `@Secret` annotation.

Marking config settings as `@Secret` will cause them to be printed as `##########` in debug.log as well as the output of `CALL dbms.listConfig`.

craigtaverner committed:
Obsfucate the LDAP system password in toString and debug.log

klaren commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/config/BaseSetting.java:
```diff
@@ -31,6 +31,7 @@
     private boolean deprecated;
     private String replacement;
     private boolean internal;
+    private boolean isSecret;
```
`secret`

klaren commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/config/BaseSetting.java:
```diff
@@ -68,6 +69,17 @@ public void setInternal( final boolean val )
+        return this.isSecret;
+    }
+
+    public void makeSecret( final boolean val )
```
`setSecret`

klaren commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/config/BaseSetting.java:
```diff
@@ -68,6 +69,17 @@ public void setInternal( final boolean val )
     }
 
+    @Override
+    public boolean isSecret()
```
`secret`

klaren commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/config/SettingGroup.java:
```diff
@@ -60,6 +60,11 @@
+    /**
+     * @return {@code true} if secret setting (should be hidden), false otherwise.
+     */
+    boolean isSecret();
```
`secret`, and this is public API so it needs to have a default implementation.

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/Config.java:
```diff
@@ -741,11 +753,16 @@ public void dump( DiagnosticsPhase phase, Logger logger )
         }
     }
 
+    private String obsfucateIfSecret( Map.Entry<String,String> param )
```
There is a `log` inside `updateDynamicSetting` that also should be obfuscated.

klaren commented on community/kernel/src/test/java/org/neo4j/kernel/configuration/ConfigTest.java:
```diff
@@ -249,16 +253,35 @@ public void shouldSetInternalParameter()
     }
 
+    @Test
+    public void shouldSetSecretParameter()
```
We should also test `updateDynamicSetting` to properly obfuscate the passwords.

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/Config.java:
```diff
@@ -82,6 +85,7 @@
     private final List<ConfigOptions> configOptions;
 
     private final Map<String,String> params = new CopyOnWriteHashMap<>(); // Read heavy workload
+    private final Set<String> secrets = new HashSet<>();
```
You don't need to populate this, you can use `settingsMap` to fetch the setting.

craigtaverner replied:
I wanted the obsfucateIfSecret method to have lowest cost, and thought a pre-populated hashmap of very small size (zero or one entry with current common cases) was the best. Otherwise we check a large collection every time.

klaren commented on community/configuration/src/main/java/org/neo4j/configuration/Secret.java:
```diff
@@ -0,0 +1,35 @@
+import java.lang.annotation.Target;
+
+/**
+ * Used to define a configuration setting as secret, which means we should not show the value in logs.
```
We should also document that this cannot be added to group settings, like connectors and SSL policies.

klaren reviewed (changes requested):
A couple of changes. 

Also, I'm not sure what level of secret we're aiming for, the values are still exposed through `org.neo4j.kernel.builtinprocs.BuiltInDbmsProcedures#listConfig` for example. And we never expose `config.toString()` to users that the does not need to be obfuscated.

craigtaverner committed:
Obsfucate secret configs in more places

* the logger passed to Configs, and called on updates
* the results from dbms.listConfigs

Even though these are admin actions, so is reading the debug.log, so
we obsfucate these cases for the same reason.

craigtaverner committed:
Removed additional HashSet since existing HashMap is good enough

klaren reviewed (approved):
Everything looks good now!

OliviaYtterbrink commented on community/configuration/src/main/java/org/neo4j/configuration/ConfigValue.java:
```diff
@@ -34,21 +34,23 @@
     private final Optional<?> value;
     private final String valueDescription;
     private final boolean internal;
+    private final boolean isSecret;
```
would prefer `secret`

OliviaYtterbrink commented on community/configuration/src/main/java/org/neo4j/configuration/ConfigValue.java:
```diff
@@ -100,6 +102,11 @@ public boolean internal()
         return internal;
     }
 
+    public boolean isSecret()
```
`secret`

OliviaYtterbrink reviewed (commented):
Looks fine, just missed one file with `isSecret` -> `secret`

craigtaverner committed:
Change isSecret to secret based on PR review
